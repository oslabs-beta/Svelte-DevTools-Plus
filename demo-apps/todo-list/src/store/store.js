import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
} from 'firebase/auth';
import { writable } from 'svelte/store';
import { auth } from '../lib/firebase/firebase';

export const authStore = writable({
  user: null,
  loading: true,
  data: {},
});

export const authHandlers = {
  signup: async (email, pass) => {
    await createUserWithEmailAndPassword(auth, email, pass);
  },
  login: async (email, pass) => {
    await signInWithEmailAndPassword(auth, email, pass);
  },
  logout: async () => {
    await signOut(auth);
  },
};
