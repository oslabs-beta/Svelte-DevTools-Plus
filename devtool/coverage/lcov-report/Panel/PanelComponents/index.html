
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for Panel/PanelComponents</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> Panel/PanelComponents</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">31.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>36/113</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/45</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">0% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>0/23</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">35% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>35/100</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file medium" data-value="ComponentInfo.tsx"><a href="ComponentInfo.tsx.html">ComponentInfo.tsx</a></td>
	<td data-value="63.63" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 63%"></div><div class="cover-empty" style="width: 37%"></div></div>
	</td>
	<td data-value="63.63" class="pct medium">63.63%</td>
	<td data-value="11" class="abs medium">7/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="63.63" class="pct medium">63.63%</td>
	<td data-value="11" class="abs medium">7/11</td>
	</tr>

<tr>
	<td class="file high" data-value="Navbar.tsx"><a href="Navbar.tsx.html">Navbar.tsx</a></td>
	<td data-value="80" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 80%"></div><div class="cover-empty" style="width: 20%"></div></div>
	</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	</tr>

<tr>
	<td class="file low" data-value="Rewinder.tsx"><a href="Rewinder.tsx.html">Rewinder.tsx</a></td>
	<td data-value="14.28" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 14%"></div><div class="cover-empty" style="width: 86%"></div></div>
	</td>
	<td data-value="14.28" class="pct low">14.28%</td>
	<td data-value="28" class="abs low">4/28</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="18.18" class="pct low">18.18%</td>
	<td data-value="22" class="abs low">4/22</td>
	</tr>

<tr>
	<td class="file low" data-value="StateModifier.tsx"><a href="StateModifier.tsx.html">StateModifier.tsx</a></td>
	<td data-value="16" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 16%"></div><div class="cover-empty" style="width: 84%"></div></div>
	</td>
	<td data-value="16" class="pct low">16%</td>
	<td data-value="25" class="abs low">4/25</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="3" class="abs low">0/3</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="19.04" class="pct low">19.04%</td>
	<td data-value="21" class="abs low">4/21</td>
	</tr>

<tr>
	<td class="file low" data-value="StateValue.tsx"><a href="StateValue.tsx.html">StateValue.tsx</a></td>
	<td data-value="45.45" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 45%"></div><div class="cover-empty" style="width: 55%"></div></div>
	</td>
	<td data-value="45.45" class="pct low">45.45%</td>
	<td data-value="11" class="abs low">5/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="20" class="abs low">0/20</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="10" class="abs low">4/10</td>
	</tr>

<tr>
	<td class="file low" data-value="TreeComponent.tsx"><a href="TreeComponent.tsx.html">TreeComponent.tsx</a></td>
	<td data-value="36.36" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 36%"></div><div class="cover-empty" style="width: 64%"></div></div>
	</td>
	<td data-value="36.36" class="pct low">36.36%</td>
	<td data-value="33" class="abs low">12/33</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="38.7" class="pct low">38.7%</td>
	<td data-value="31" class="abs low">12/31</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-10-26T22:15:09.485Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    