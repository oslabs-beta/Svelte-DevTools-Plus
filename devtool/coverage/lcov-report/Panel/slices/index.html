
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for Panel/slices</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> Panel/slices</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">75.75% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>25/33</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">100% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>0/0</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">62.5% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>5/8</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">75.75% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>25/33</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input oninput="onInput()" type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line medium'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="currentSnapshotSlice.tsx"><a href="currentSnapshotSlice.tsx.html">currentSnapshotSlice.tsx</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	</tr>

<tr>
	<td class="file medium" data-value="highlightedComponentSlice.ts"><a href="highlightedComponentSlice.ts.html">highlightedComponentSlice.ts</a></td>
	<td data-value="50" class="pic medium">
	<div class="chart"><div class="cover-fill" style="width: 50%"></div><div class="cover-empty" style="width: 50%"></div></div>
	</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="14" class="abs medium">7/14</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="50" class="pct medium">50%</td>
	<td data-value="14" class="abs medium">7/14</td>
	</tr>

<tr>
	<td class="file high" data-value="treeHistorySlice.tsx"><a href="treeHistorySlice.tsx.html">treeHistorySlice.tsx</a></td>
	<td data-value="90" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 90%"></div><div class="cover-empty" style="width: 10%"></div></div>
	</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="10" class="abs high">9/10</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="66.66" class="pct medium">66.66%</td>
	<td data-value="3" class="abs medium">2/3</td>
	<td data-value="90" class="pct high">90%</td>
	<td data-value="10" class="abs high">9/10</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2023-10-31T00:03:24.256Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    