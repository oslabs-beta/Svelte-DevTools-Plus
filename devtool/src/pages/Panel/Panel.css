html,
body {
  height: 100%;
  margin: 0;
  padding: 0;
}

body {
  --main-bg-color: #444e62;
  --rewinder-height: 2.5rem;
  background-color: var(--main-bg-color);
}

.container {
  color: #eeeeee;
}

.pane {
  display: flex;
  flex-direction: column;
  height: 100%;
  background-color: var(--main-bg-color);
  font-family: 'Source Sans Pro', sans-serif;
}

.nav-option {
  margin-right: 1rem;
  margin-left: 1rem;
}

.nav-option:hover {
  background-color: rgba(50, 50, 50, 0.25);
}

.split {
  display: flex;
  flex-direction: row;
}

.gutter {
  background-color: #eee;
  background-repeat: no-repeat;
  background-position: 50%;
}

.gutter.gutter-horizontal {
  background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAeCAYAAADkftS9AAAAIklEQVQoU2M4c+bMfxAGAgYYmwGrIIiDjrELjpo5aiZeMwF+yNnOs5KSvgAAAABJRU5ErkJggg==');
  cursor: col-resize;
}

.nav {
  display: flex;
  align-items: stretch;
}

.nav ul {
  display: flex;
  /* gap: 1rem; */
  list-style: none;
  margin: 0;
  padding: 0;
}

.nav a {
  color: inherit;
  text-decoration: none;
  display: flex;
  align-items: stretch;
  justify-content: center;
}

.nav li {
  color: rgb(28, 195, 221);
}

.nav-link {
  padding: 0.5rem;
  cursor: pointer;
  border-radius: 0.25rem;
}

.nav-link:active {
  background-color: rgb(55, 68, 114);
}
.nav-link:hover {
  background-color: rgb(114, 137, 218);
}

.page-content {
  height: calc(100% - var(--rewinder-height));
}

#treeWrapper,
.container,
.split {
  height: 100%;
}

.container {
  height: 100%;
}

.step-page-gap {
  height: 2rem;
}

header {
  height: 3.25rem;
  max-height: 3.25rem;
  border-bottom: #eeeeee solid 2px;
  display: flex;
  justify-content: center;
  align-items: center;
}

#no-data-error {
  text-align: center;
}

.pane-content {
  overflow: overlay;
  height: calc(100% - var(--rewinder-height));
}

#root-container {
  padding-top: 2rem;
  padding-bottom: 2rem;
}
