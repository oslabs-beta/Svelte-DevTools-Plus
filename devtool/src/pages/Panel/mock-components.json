{
  "component": "Root",
  "componentState": null,
  "componentProps": null,
  "children": [
    {
      "component": "App",
      "componentState": null,
      "componentProps": null,
      "children": [
        {
          "component": "Board",
          "componentState": {
            "boardState": [
              ["O", "X", "X"],
              ["-", "-", "-"],
              ["-", "-", "O"]
            ],
            "turnCount": 0,
            "turn": "X",
            "gameOver": false,
            "message": ""
          },
          "componentProps": null,
          "children": [
            {
              "component": "Row",
              "componentState": null,
              "componentProps": {
                "rowState": ["O", "X", "X"],
                "handleClick": "function(){console.log('clicked');}",
                "rowIndex": 0
              },
              "children": [
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "O",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 0
                  },
                  "children": null
                },
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "X",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 1
                  },
                  "children": null
                },
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "X",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 2
                  },
                  "children": null
                }
              ]
            },
            {
              "component": "Row",
              "componentState": null,
              "componentProps": {
                "rowState": ["-", "-", "-"],
                "handleClick": "function(){console.log('clicked');}",
                "rowIndex": 1
              },
              "children": [
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "-",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 0
                  },
                  "children": null
                },
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "-",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 1
                  },
                  "children": null
                },
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "-",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 2
                  },
                  "children": null
                }
              ]
            },
            {
              "component": "Row",
              "componentState": null,
              "componentProps": {
                "rowState": ["-", "-", "O"],
                "handleClick": "function(){console.log('clicked');}",
                "rowIndex": 2
              },
              "children": [
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "-",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 0
                  },
                  "children": null
                },
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "-",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 1
                  },
                  "children": null
                },
                {
                  "component": "Box",
                  "componentState": null,
                  "componentProps": {
                    "boxState": "O",
                    "handleClick": "function(){console.log('clicked');}",
                    "boxIndex": 2
                  },
                  "children": null
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}