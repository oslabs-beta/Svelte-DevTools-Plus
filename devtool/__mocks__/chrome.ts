import { Component } from '../src/pages/Panel/slices/highlightedComponentSlice';
import initialData from './mockData';

let data = JSON.parse(JSON.stringify(initialData));

type ChromeMessageListener = (message: any) => void;

interface MockRuntime {
  onMessage: {
    addListener: (callback: ChromeMessageListener) => void;
    _triggerMessage: (message: any) => void;
  };
  sendMessage: (message: MockMessageType) => void;
}

interface MockMessageType {
  message:
    | 'getRootComponent'
    | 'getSvelteVersion'
    | 'handleClosedPanel'
    | 'injectState'
    | 'injectSnapshot';
  snapshot?: Component;
  componentId?: number;
  newState?: {
    [stateKey: string]: number | string;
  };
}

interface MockTabs {
  query: (queryInfo: any) => any[];
  sendMessage: (tabId: number, message: MockMessageType) => void;
}

export interface MockChrome {
  runtime: MockRuntime;
  tabs: MockTabs;
  clearListeners: () => void;
  resetMockData: () => void;
  sendEmptyDataOnNextRequest: () => void;
}

let listeners: ChromeMessageListener[] = [];
let _sendEmptyDataOnNextRequest = false;

function updateState(id: number | undefined, newState: any): Boolean {
  function helper(component: any): Boolean {
    if (component.id === id) {
      for (let i = 0; i < component.detail.ctx.length; i++) {
        const state = component.detail.ctx[i];
        if (state.key === Object.keys(newState)[0]) {
          const value = Object.values(newState)[0];
          if (state.value === value) return false;
          state.value = value;
          return true;
        }
      }
      return false;
    }
    for (let i = 0; i < component.children.length; i++) {
      const child = component.children[i];
      return helper(child);
    }
    return false;
  }
  return helper(data);
}

const chrome: MockChrome = {
  runtime: {
    onMessage: {
      addListener: (callback) => {
        listeners.push(callback);
      },
      _triggerMessage: (message) => {
        listeners.forEach((callback) => callback(message));
      },
    },
    sendMessage: function (message) {},
  },
  tabs: {
    query: () => {
      return [{ id: 0 }];
    },
    sendMessage: (tabId, request) => {
      switch (request.message) {
        case 'getRootComponent':
          {
            let message: {};
            if (_sendEmptyDataOnNextRequest) {
              message = {
                type: 'returnRootComponent',
                rootComponent: undefined,
              };
              _sendEmptyDataOnNextRequest = false;
            } else {
              message = {
                type: 'returnRootComponent',
                rootComponent: JSON.parse(JSON.stringify(data)),
              };
            }
            listeners.forEach((f) => f(message));
          }
          break;
        case 'getSvelteVersion':
          {
          }
          break;
        case 'handleClosedPanel':
          {
          }
          break;
        case 'injectState':
          {
            const stateUpdated = updateState(
              request.componentId,
              request.newState
            ); // Don't send an update if nothing was changed
            // This causes problems
            if (!stateUpdated) return;
            const message = {
              type: 'updateRootComponent',
              rootComponent: JSON.parse(JSON.stringify(data)),
            };
            listeners.forEach((f) => f(message));
          }
          break;
        case 'injectSnapshot':
          {
            if (!request.snapshot) return;
            data = JSON.parse(JSON.stringify(request.snapshot));
            const message = {
              type: 'returnTempRoot',
              rootComponent: JSON.parse(JSON.stringify(data)),
            };
            listeners.forEach((f) => f(message));
          }
          break;
      }
    },
  },
  clearListeners: function () {
    listeners = [];
  },
  resetMockData: function () {
    data = JSON.parse(JSON.stringify(initialData));
  },
  sendEmptyDataOnNextRequest: function () {
    _sendEmptyDataOnNextRequest = true;
  },
};

export default chrome;
