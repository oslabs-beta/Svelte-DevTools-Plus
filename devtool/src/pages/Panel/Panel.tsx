import { useEffect } from 'react';
import './Panel.css';
import Split from 'react-split';
import ComponentInfo from './PanelComponents/ComponentInfo';
import Navbar from './PanelComponents/Navbar';
import { Routes, Route, useNavigate } from 'react-router-dom';
import TreePage from './PanelPages/TreePage';
import StepPage from './PanelPages/StepPage';
import { Component } from './slices/highlightedComponentSlice';
import { useSelector } from 'react-redux';
import { selectCurrentSnapshot } from './slices/currentSnapshotSlice';
import { useDispatch } from 'react-redux';
import { TreeHistory, selectTreeHistory } from './slices/treeHistorySlice';
import Rewinder from './PanelComponents/Rewinder';

export interface ComponentPageProps {
  rootComponentData: Component;
}

// Let the rest of the extension know that the panel is closed
// so it won't try and send messages to it
window.addEventListener('beforeunload', function () {
  chrome.runtime.sendMessage({ message: 'handleClosedPanel' });
});

function Panel() {
  const currentSnapshot = useSelector(selectCurrentSnapshot);
  const treeHistory: TreeHistory = useSelector(selectTreeHistory);
  const navigate = useNavigate();
  const dispatch = useDispatch();

  // Navigate to the root directory on page load
  useEffect(() => {
    navigate('/');
  }, []);

  useEffect(() => {
    async function setUpPanel() {
      const [tab] = await chrome.tabs.query({
        active: true,
        lastFocusedWindow: true,
      });
      chrome.tabs.sendMessage(tab.id!, { message: 'getRootComponent' });
    }
    setUpPanel();
    // I only want to add a listener once, so it goes in the onMount useEffect
    // Listens for response from ContentScriptIsolated. This is where we
    // get the current tab's root component, and process updates
    chrome.runtime.onMessage.addListener(function (message: any) {
      if (message.type === 'updateRootComponent') {
        const rootComponent = message.rootComponent;
        if (rootComponent) {
          createAndSaveNewSnapshot(rootComponent);
        }
      } else if (message.type === 'returnRootComponent') {
        const rootComponent = message.rootComponent;
        if (rootComponent) {
          createAndSaveNewSnapshot(rootComponent);
        } else {
          console.log('Error getting component data');
        }
        // For use after rewinding
      } else if (message.type === 'returnTempRoot') {
        const tempRoot = message.rootComponent;
        // set the tempRoot as the current snapshot without saving it
        dispatch({
          type: 'currentSnapshot/setCurrentSnapshot',
          payload: {
            rootComponent: tempRoot,
          },
        });
      }
    });
  }, []);

  function createAndSaveNewSnapshot(newRootComponent: Component) {
    dispatch({
      type: 'currentSnapshot/setCurrentSnapshot',
      payload: {
        rootComponent: newRootComponent,
      },
    });
    dispatch({
      type: 'treeHistory/addNewSnapshot',
      payload: {
        newSnapshot: newRootComponent,
      },
    });
  }

  // Takes the snapshot in treeHistory[snapshotIndex] and injects state with all
  // of its component's state. Then the content script returns a "temp root component"
  // which gets displayed to the screen without saving a new snapshot
  async function changeSnapshot(snapshotIndex: number) {
    const [tab] = await chrome.tabs.query({
      active: true,
      lastFocusedWindow: true,
    });
    console.log('butts', treeHistory.treeHistory);
    chrome.tabs.sendMessage(tab.id!, {
      message: 'injectSnapshot',
      snapshot: treeHistory.treeHistory[snapshotIndex],
    });
  }

  function clearSnapshotHistory() {
    dispatch({
      type: 'treeHistory/deleteAllSnapshots',
    });
  }

  return (
    <div className="container" data-testid="panel">
      <div id="content">
        <Split className="split">
          <div className="pane">
            <header>
              <Navbar />
            </header>
            <Routes>
              <Route
                path="/"
                element={
                  <StepPage
                    rootComponentData={currentSnapshot.rootComponent!}
                  />
                }
              />
              <Route
                path="/tree"
                element={
                  <TreePage
                    rootComponentData={currentSnapshot.rootComponent!}
                  />
                }
              />
            </Routes>
          </div>
          <div className="pane">
            <ComponentInfo />
          </div>
        </Split>
      </div>
      <Rewinder
        changeSnapshot={changeSnapshot}
        numberOfSnapshots={treeHistory.treeHistory.length}
        clearSnapshotHistory={clearSnapshotHistory}
      />
    </div>
  );
}

export default Panel;
