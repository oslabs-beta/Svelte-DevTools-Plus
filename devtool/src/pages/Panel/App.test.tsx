/**
 * @jest-environment jsdom
 */
// import { describe, expect, test } from '@jest/globals';
import { render, screen, cleanup } from '@testing-library/react';
import Panel from './Panel';
import { Provider } from 'react-redux';
import { store } from './store';
import { BrowserRouter } from 'react-router-dom';
import { jest } from '@jest/globals';
import chrome from '../../../__mocks__/chrome';
import { act } from 'react-dom/test-utils';

jest.mock('chrome');

describe('test tests', function () {
  beforeEach(() => {
    // I get a "chrome is not defined" error for some reason without this ???
    chrome;
  });

  it('testing works', async () => {
    await act( async () => render(
      <Provider store={store}>
        <BrowserRouter>
          <Panel />
        </BrowserRouter>
      </Provider>
    ));
    const element = screen.getByTestId('root-component');
    console.log(element);
  });
});

// describe('sum module', () => {
//   test('adds 1 + 2 to equal 3', () => {
//     expect(1 + 2).toBe(3);
//   });
// });
