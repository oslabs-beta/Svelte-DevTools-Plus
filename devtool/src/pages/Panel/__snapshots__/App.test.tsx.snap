// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel tests Matches the snapshot 1`] = `
<div
  class="container"
  data-testid="panel"
>
  <div
    id="content"
  >
    <div
      class="split"
    >
      <div
        class="pane"
        style="width: calc(50% - 5px);"
      >
        <header>
          <nav
            class="nav"
          >
            <ul>
              <li
                class="step"
              >
                <a
                  href="/"
                >
                  Step
                </a>
              </li>
              <li
                class="tree"
              >
                <a
                  data-testid="tree-link"
                  href="/tree"
                >
                  Tree
                </a>
              </li>
            </ul>
          </nav>
        </header>
        <div
          class="pane"
        >
          <div
            class="step-page-gap"
          />
          <div
            data-testid="root-container"
          >
            <div
              tabindex="0"
            >
              <div>
                <div
                  class="tree-component"
                >
                  <div
                    data-testid="expand-button-App"
                  >
                    <img
                      class="expand-button"
                      src="test-file-stub"
                    />
                  </div>
                  <div
                    class="tree-component-bar"
                    data-testid="component-button-App"
                  >
                    &lt;
                    <span
                      class="component-name"
                    >
                      App
                    </span>
                    /&gt;
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="gutter gutter-horizontal"
        style="width: 10px;"
      />
      <div
        class="pane"
        style="width: calc(50% - 5px);"
      >
        <div
          class="pane"
          id="component-info"
        >
          <header>
            <h2 />
          </header>
          <h3>
            State
          </h3>
          <h3>
            Props
          </h3>
        </div>
      </div>
    </div>
  </div>
  <div
    id="rewinder"
  >
    <button
      id="clear-button"
    >
      Clear
    </button>
    <div
      id="slider-container"
    >
      <span
        class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium css-187mznn-MuiSlider-root"
      >
        <span
          class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style="left: 0%; width: 100%;"
        />
        <span
          class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index="0"
          style="left: 0%;"
        />
        <span
          class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index="1"
          style="left: 100%;"
        />
        <span
          class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-eg0mwd-MuiSlider-thumb"
          data-index="0"
          style="left: 100%;"
        >
          <input
            aria-label="Temperature"
            aria-orientation="horizontal"
            aria-valuemax="2"
            aria-valuemin="1"
            aria-valuenow="2"
            data-index="0"
            max="2"
            min="1"
            step="1"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="2"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                2
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
    <button
      data-testid="rewind-button"
      id="back-button"
    >
      &lt;&lt;
    </button>
    <button
      data-testid="revert-button"
      id="forward-button"
    >
      &gt;&gt;
    </button>
  </div>
</div>
`;
